### PR TITLE
Add codespell pre-commit hook, fix typos in CCCL

### PR DESCRIPTION
**Pull request by bdice**
## Description

This PR adds `codespell` to the list of pre-commit hooks enforced in CCCL.

We have adopted `codespell` in RAPIDS and found its false-positive rate to be very low.

I have three commits in this PR.

The first sets up the `codespell` configuration. I created a list of words that are false positives (or files to skip) in this first commit, by manually reviewing the output of `codespell -i 3`.

The second commit includes the output of `codespell -w` with those exclusions applied.

The third commit is contains some manual fixes for ambiguous spellings, fixes some issues in the config, and adds notes on how to run `codespell -i 3 -w -H` when using the tool outside of pre-commit.

If desired, I can open separate PRs to fix typos in each subproject, and then add the pre-commit hook at the end (or gradually add files to the scope of files it handles).

## Checklist
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by bdice** [comment]
I need to update the last commit with manual fixes for the ambiguous words. Then this will be ready for review!

**Review by miscco — APPROVED** [review]
This is amazing

**Comment by miscco** [comment]
pre-commit.ci autofix

File: thrust/testing/for_each.cu [hunk 1]
```diff
@@ -318,7 +318,11 @@ void TestForEachNWithLargeTypes()
   _TestForEachNWithLargeTypes<int, 256>();
   _TestForEachNWithLargeTypes<int, 512>();
 
+<<<<<<< Updated upstream
   // XXX parallel_for does't support large types
+=======
+  // XXX parallel_for doesn't support large types
+>>>>>>> Stashed changes
```
Comment (miscco) [inline thrust/testing/for_each.cu:325]:
```suggestion
#include <cuda/__cccl_config> 
```

**Comment by bdice** [comment]
Alright, I force-pushed some changes to the last "manual" commit that fix the ambiguous errors and update the configs a little bit. This should be ready for review.

**Review by rwgk — APPROVED** [review]
Thanks a ton for doing this!

**Pull request merged**